### ISSUE TITLE
DML execution provider doesn't load model while CUDA,CPU execution provider does

### ISSUE THREAD

**Issue by Mut1nyJD** [microsoft/onnxruntime#6457]
**Describe the bug**

I've got a Opset11 model but get an exception when using the DML execution provider. I followed the guidelines as well disabled all optimizations set ExecutionMode to SEQUENTIAL and Disable MemPatterns.
Still when I try to Create a session with this specific model I'll get an exception:

Exception during initialization: 
D:\3\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\FusedGraphKernel.cpp(64)\onnxruntime.dll!00007FF806AF4FBD: (caller: 00007FF806AF4578) Exception(1) tid(9770) 80004002 No such interface supported

I am using latest NuGet package 1.6.0 which also installed DirectML 1.4.0

**Urgency**
Medium I can use CPU or CUDA for now but would like to see this solved so I can switch to DirectML and do not need to include multiple versions 

**System information**
Windows 10 
ONNX Runtime installed from binary see above
1.6.0
Visual Studio 2019
Nivida Mobile GTX1060ti 6GB VRAM 

**To Reproduce**

```
const OrtApi* g_ort = OrtGetApiBase()->GetApi(ORT_API_VERSION);

void CheckStatus(OrtStatus* status)
{
	if (status != NULL) {
		const char* msg = g_ort->GetErrorMessage(status);
		fprintf(stderr, "%s\n", msg);
		g_ort->ReleaseStatus(status);
		exit(1);
	}
}
	OrtSession* m_session = NULL;
	OrtSessionOptions* m_session_opts = NULL;
	OrtAllocator* m_allocator = NULL;
	OrtMemoryInfo* m_memory_info = NULL;

	if (g_env == NULL)
		CheckStatus(g_ort->CreateEnv(ORT_LOGGING_LEVEL_ERROR, "test", &g_env));
	g_ort->SetSessionExecutionMode(m_session_opts, ORT_SEQUENTIAL);
	g_ort->SetIntraOpNumThreads(m_session_opts, 8);
	g_ort->SetInterOpNumThreads(m_session_opts, 8);
	g_ort->DisableMemPattern(m_session_opts);
	g_ort->SetSessionExecutionMode(m_session_opts, ExecutionMode::ORT_SEQUENTIAL);
	OrtSessionOptionsAppendExecutionProvider_DML(m_session_opts, 1);
	CheckStatus(g_ort->GetAllocatorWithDefaultOptions(&m_allocator));
	CheckStatus(g_ort->CreateCpuMemoryInfo(OrtArenaAllocator, OrtMemTypeDefault, &m_memory_info));
       CheckStatus(g_ort->CreateSession(g_env, "mymodel.onnx", m_session_opts, &m_session));

```
Too big but can provide if needed via external link

**Expected behavior**
Not to throw an exception

**Comment by fdwr** [comment]
@Mut1nyJD - can you link the model (removing any confidential weights or proprietary logic)? Then we can try to repro this with WinML to see if it's unique to ORT usage (more likely it happens with both, inside ORT's DirectML execution provider). TY.

**Comment by Mut1nyJD** [comment]
@fdwr 
> @Mut1nyJD - can you link the model (removing any confidential weights or proprietary logic)? Then we can try to repro this with WinML to see if it's unique to ORT usage (more likely it happens with both, inside ORT's DirectML execution provider). TY.

Sure thing, please find download link below

https://cloud.degoo.com/share/OSztmu1W2nbGkg

**Comment by fdwr** [comment]
Thanks. Downloaded the model "yolov5_face.onnx".

🤔 The model runs successfully with WinMlRunner.exe (from here https://github.com/Microsoft/Windows-Machine-Learning/tree/master/Tools/WinMLRunner) using ORT master bits along with DirectML 1.4.1 (not published yet, just internally). So I'll check against DirectML 1.4.0 and ORT 1.6 too to see if I can repro it.

```
C:\>WinMLRunner.exe -gpu -model "yolov5_face onnxruntime github issue 6457.onnx"

Created LearningModelDevice with GPU: NVIDIA Quadro P400
Loading model (path = D:\ai\models\broken\yolov5_face onnxruntime github issue 6457.onnx)...
=================================================================
Name: torch-jit-export
Author: pytorch
Version: 0
Domain:
Description:
Path: D:\ai\models\broken\yolov5_face onnxruntime github issue 6457.onnx
Support FP16: false

Input Feature Info:
Name: images
Feature Kind: Float

Output Feature Info:
Name: output
Feature Kind: Float

Name: 566
Feature Kind: Float

Name: 585
Feature Kind: Float

=================================================================

Binding (device = GPU, iteration = 1, inputBinding = CPU, inputDataType = Tensor, deviceCreationLocation = WinML)...[SUCCESS]
Evaluating (device = GPU, iteration = 1, inputBinding = CPU, inputDataType = Tensor, deviceCreationLocation = WinML)...[SUCCESS]
```

**Comment by Mut1nyJD** [comment]
Any update on this? As it turns out there are lots of models that do not work anything that is basically not just a pure image classification model I'll get the same error, including the FasterRCN and MaskRCN models from the official ONNX model zoo.

**Comment by Mut1nyJD** [comment]
Still seeing the same issue with Version 1.7.0 DML nuGet package I am wondering if there is something wrong on my end?

About 75% of the models on the official model zoo on www.onnx.ai don't load with the DML ExecutionProvider active.

The only ones I managed to get running is the MobileNetV2 and  candy.onnx that is referenced in the fns_candy_style_transfer sample code. 

Anything else does not load making this Execution Provider pretty useless tbh.

**Comment by fdwr** [comment]
Can you confirm that the newer DirectML.dll is being loaded by your executable? (e.g. in the Visual Studio debugger Modules window). If the DLL is copied alongside your exe, then the new DLL should be found. Otherwise ORT is likely trying to DirectML.dll from the OS in C:\Windows\System32\* (which is older).

**Comment by fdwr** [comment]
> it works now
Yay.

>The other use case I had been trying before was where my inference code also is inside its own .dll that gets loaded by another executable

@Mut1nyJD : Yeah, LoadLibrary calls favor the process (.exe) path, not that of the loading DLL. 🤔 I'm not sure what to suggest in that case, except maybe have your DLL call `AddDllDirectory` before initializing the ORT and the DML EP (and to be clean, `RemoveDllDirectory` afterward). If the ONNX model zoo works correctly now (albeit slowly) for you, is it safe to close the issue?

**Comment by Mut1nyJD** [comment]
@fdwr : Yeah I spend hours reading through assembly manifest documentation to see if there is something I can do because frankly I don't understand what is happening since it does work with onnxruntime.dll it is loaded from the same directory as my .dll but the dependent DirectML.dll is not. In the end I did something similar to your suggestion so I added a SetDllDirectory call in my .dll a bit of a hack but oh well. 

Anyway thank you for your help I will close this as not only does it work with the model zoo ones but also with the original model I was initially filling the issue about.
